boards: ti_am62x_sk: Add support for Cortex M4 core in TI AM62x SK EVM board

KarthikL1729 opened the pull request:
Add support files for the Texas Instruments AM62x SK board, for the Cortex M4F core. hello_world sample test build successfully with these configurations and it has been tested by remoteproc loading on a board. The board prints the output to the uart console.

Signed-off-by: L Lakshmanan <l-lakshmanan@ti.com>

gramsay0 reviewed (commented):
Good work, and good idea using the k3 naming.
The AM62x Cortex-A53 support probably should be renamed Sitara -> k3 to encompass the wider SoC family as well

KarthikL1729 commented on boards/arm/ti_am62x_sk/ti_am62x_sk_defconfig:
```diff
@@ -0,0 +1,25 @@
+# Cache Options
+CONFIG_CACHE_MANAGEMENT=y
+CONFIG_DCACHE_LINE_SIZE_DETECT=y
+CONFIG_ICACHE_LINE_SIZE_DETECT=y
```
Yes, it was added accidentally, resolved now.

KarthikL1729 commented on boards/arm/ti_am62x_sk/Kconfig.board:
```diff
@@ -0,0 +1,9 @@
+# Texas Instruments Sitara AM62x-SK EVM
```
Have changed all file names and folder names, as well as defines to show m4. Tested and verified build after changes.

KarthikL1729 commented on dts/arm/ti/am62x_sk.dtsi:
```diff
@@ -0,0 +1,52 @@
+	soc {
+	};
+	model = "";
+	compatible = "";
```
I had added them to indicate possible addition of peripherals, but have removed them now as they are not essential.

KarthikL1729 commented on boards/arm/ti_am62x_sk/doc/index.rst:
```diff
@@ -0,0 +1,79 @@
+
+`UniFlash`_ command line utility is used to program the flash memory. Only
+elf loading is currently supported.
+
```
I am trying to allow the board to be flashed without any vendor specific tools, and so have removed this. Will keep you updated!

gramsay0 replied:
Sorry my comment was misleading, and not specifically targeted at these lines of code.
A vendor specific command line utility like "UniFlash" is fine.
My preference is only to avoid requiring vendor IDEs, which you had already done.
All I suggest is to provide clear steps that will allow others to easily program the board

DhruvaG2000 commented on boards/arm/ti_am62x_sk_m4/doc/index.rst:
```diff
@@ -0,0 +1,74 @@
+
+The AM62x-SK board configuration is used by Zephyr applications that run on
+the TI AM62x platform. The board configuration provides support for the ARM
+Cortex-M4F CPU and the following devices:
```
Better to say "Cortex-M4F microcontroller/MCU core" than Cortex-M4F "CPU"?

KarthikL1729 replied:
Made the change. Thanks for the suggestion!

DhruvaG2000 commented on boards/arm/ti_am62x_sk_m4/doc/index.rst:
```diff
@@ -0,0 +1,74 @@
+the TI AM62x platform. The board configuration provides support for the ARM
+Cortex-M4F CPU and the following devices:
+
+- Nested Vector Interrupt Controller (NVIC)
```
I don't think this comes under "having support for following devices".

NVIC is used to handle all the interrupts and exceptions that Cortex-M supports, I'd rather call it as "supported features" or something on those lines?

KarthikL1729 replied:
Changed it to features from devices.

DhruvaG2000 commented on boards/arm/ti_am62x_sk_m4/doc/index.rst:
```diff
@@ -0,0 +1,74 @@
+
+Hardware
+********
+
```
Please add a line here something on the lines of:

The AM62x SoC has 1 MCU domain with the following specs:

KarthikL1729 replied:
Have added the line now.

DhruvaG2000 commented on boards/arm/ti_am62x_sk_m4/ti_am62x_sk_m4_defconfig:
```diff
@@ -0,0 +1,20 @@
+CONFIG_SEMIHOST=y
+
+# Zephyr Kernel Configuration
+CONFIG_XIP=n
```
any particular reason to set `CONFIG_XIP=n` ? Can you try and remove this line and check if it works?

KarthikL1729 replied:
> I can see that you have added `soc/arm/ti_k3` [ca7c3a2](https://github.com/zephyrproject-rtos/zephyr/commit/ca7c3a23fb9d1f269ecfd43e701236195f2e72f8) and then in [a72e4e0](https://github.com/zephyrproject-rtos/zephyr/commit/a72e4e0a2fc2183d94bec62a2905a95f6e5ed530) you are editing some of the files you yourself had added in the earlier commit. This is in general a practise to **avoid**, especially since you are adding completely new files in your PR, you can just add them with the necessary changes at once, not in 2 seperate commits unless there's a very good reason for doing so? I think that is one reason also why compliance checks maybe failing, although I maybe wrong about that.

Yes, I observed that too, have rebased and fixed it now.

KarthikL1729 replied:
> any particular reason to set `CONFIG_XIP=n` ? Can you try and remove this line and check if it works?

Right now, the build fails as it overflows, due to the ram being very small. Will attempt after adding access to more memory.

DhruvaG2000 commented on boards/arm/ti_am62x_sk_m4/ti_am62x_sk_m4.dts:
```diff
@@ -0,0 +1,29 @@
+	cpus {
+		cpu@0 {
+			status = "okay";
+			clock-frequency = < 400000000 >;
```
No need of spaces in <> 
~< 400000000 >~

`<400000000>;` should be fine

DhruvaG2000 commented on soc/arm/ti_k3/am62x_m4/soc.h:
```diff
@@ -0,0 +1,5 @@
+ * Copyright (c) 2023 L Lakshmanan
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
If we are going to leave this file blank do we really need it?
What happens if you remove `#include <soc.h>` from `soc.c` ?

KarthikL1729 replied:
I have kept the file for future additions to the soc files. As of now though, it builds fine without `soc.h`.

DhruvaG2000 reviewed (changes requested):
Hi,
Thanks for the PR, it looks good but I had some minor suggestions. Please take a look and address those comments.
Welcome to Zephyr and congrats on creating your first PR!

DhruvaG2000 commented:
I can see that you have added `soc/arm/ti_k3` https://github.com/zephyrproject-rtos/zephyr/pull/58722/commits/ca7c3a23fb9d1f269ecfd43e701236195f2e72f8 and then in a72e4e0a2fc2183d94bec62a2905a95f6e5ed530  you are editing some of the files you yourself had added in the earlier commit.
This is in general a practise to **avoid**, especially since you are adding completely new files in your PR, you can just add them with the necessary changes at once, not in 2 seperate commits unless there's a very good reason for doing so?
I think that is one reason also why compliance checks maybe failing, although I maybe wrong about that.

vaishnavachath commented:
@KarthikL1729 ,
Thank you for your Pull request, Nice work! I have provided some inline comments.
Also would you prefer getting this PR accepted once UART console is working? right now you need a debugger to view the semihosted debug console and very limited peripherals(Systick, NVIC) are supported, also the flashing steps are custom/needs Uniflash? I would recommend getting UART working before merging this PR  so as to have a better developer/user experience for future developers working on this platform, but am open to suggestions if this needs to merged sooner[after merge window opens].

vaishnavachath commented on boards/arm/ti_am62x_sk_m4/Kconfig.board:
```diff
@@ -0,0 +1,9 @@
+# Texas Instruments Sitara AM62x-SK-M4 EVM
+#
+# Copyright (c) 2023 L Lakshmanan
+#
```
Here and other places, use a TI copyright statement as this is  part of an official TI project.
`Copyright (C) 2023 Texas Instruments Incorporated`

grantramsay replied:
@vaishnavachath isn't that only the case if the contributor is employed by TI?

DhruvaG2000 replied:
> @vaishnavachath isn't that only the case if the contributor is employed by TI?

In this case, Lakshmanan (the contributor) is interning at TI and you can see all his commits are signed off by @ti.com domain.

Hence  I believe Vaishnav has made the above suggestion.

KarthikL1729 replied:
Fixed in all files now.

vaishnavachath commented on boards/arm/ti_am62x_sk_m4/doc/index.rst:
```diff
@@ -0,0 +1,75 @@
+
+   boards/arm/ti_am62x_sk_m4/ti_am62x_sk_m4_defconfig
+
+
```
Are you providing any instructions on how to load and run any example, since it requires custom steps those should be included in the documentation.
Also can you check a preview of this README and make other corrections : https://github.com/zephyrproject-rtos/zephyr/blob/b595082c98ebbcfb6d457742d30c973e27c57bcb/boards/arm/ti_am62x_sk_m4/doc/index.rst

KarthikL1729 replied:
Added a section on flashing, mentioning that it is only possible through CCS as of now.

vaishnavachath commented on boards/arm/ti_am62x_sk_m4/ti_am62x_sk_m4.dts:
```diff
@@ -0,0 +1,29 @@
+	};
+
+	dram0: memory@82000000 {
+		reg = <0x82000000 DT_SIZE_M(1)>;
```
Is this DRAM region untouched by other components? is this always fixed?

KarthikL1729 replied:
It is untouched by other components as of now. I left it as a placeholder for now and it can be used when we need to add dram access.

vaishnavachath commented on dts/arm/ti/am62x_sk_m4.dtsi:
```diff
@@ -0,0 +1,47 @@
+		#size-cells = <0>;
+
+		cpu0: cpu@0 {
+				device_type = "cpu";
```
Fix indentation.

KarthikL1729 replied:
Fixed.

KarthikL1729 commented:
> @KarthikL1729 , Thank you for your Pull request, Nice work! I have provided some inline comments. Also would you prefer getting this PR accepted once UART console is working? right now you need a debugger to view the semihosted debug console and very limited peripherals(Systick, NVIC) are supported, also the flashing steps are custom/needs Uniflash? I would recommend getting UART working before merging this PR so as to have a better developer/user experience for future developers working on this platform, but am open to suggestions if this needs to merged sooner[after merge window opens].

I think it would be preferable to merge this PR as it is, as UART for the M4F core requires an address translation module to be implemented (RAT) which is TI specific, along with some other configurations, and doing that will take some time. This could be a basic platform for now and hence merging it would be preferred.

vaishnavachath commented:
> > @KarthikL1729 , Thank you for your Pull request, Nice work! I have provided some inline comments. Also would you prefer getting this PR accepted once UART console is working? right now you need a debugger to view the semihosted debug console and very limited peripherals(Systick, NVIC) are supported, also the flashing steps are custom/needs Uniflash? I would recommend getting UART working before merging this PR so as to have a better developer/user experience for future developers working on this platform, but am open to suggestions if this needs to merged sooner[after merge window opens].
> 
> I think it would be preferable to merge this PR as it is, as UART for the M4F core requires an address translation module to be implemented (RAT) which is TI specific, along with some other configurations, and doing that will take some time. This could be a basic platform for now and hence merging it would be preferred.

@KarthikL1729  without the basic support needed to enable a peripheral and no UART support, upstreaming this board support is not very useful in my opinion and also steps  for loading the FW required you to install CCS which again is not the best path if you need to enable zephyr support for this target. Let's work together to get the support to a better state.

vaishnavachath commented:
Marking as DNM as few change requests needs to addressed, requesting inputs from other maintainers on the RAT/External Memory Translation Unit implementation.

CC: @carlocaione

carlocaione commented:
Alright, I think we need a new MMIO API or something like that. Basically what you did https://github.com/zephyrproject-rtos/zephyr/pull/58722/commits/6390a2dcbc8ad923ae049c74eae4904a26cbdeee can be acceptable but `z_get_local_addr` must be part of a proper MMIO API so that we can use different backends / implementations if needed (RAT being one of these) and in general that must be cleaned up a bit to seamlessly support the regular MMU backend and the new MMIO API.

@vaishnavachath @KarthikL1729 this new API / subsystem should be pushed as standalone PR if possible.

@dcpleung what do you think?

vaishnavachath commented:
> Alright, I think we need a new MMIO API or something like that. Basically what you did [6390a2d](https://github.com/zephyrproject-rtos/zephyr/commit/6390a2dcbc8ad923ae049c74eae4904a26cbdeee) can be acceptable but `z_get_local_addr` must be part of a proper MMIO API so that we can use different backends / implementations if needed (RAT being one of these) and in general that must be cleaned up a bit to seamlessly support the regular MMU backend and the new MMIO API.
> 
> @vaishnavachath @KarthikL1729 this new API / subsystem should be pushed as standalone PR if possible.
> 
> @dcpleung what do you think?

Thank you for the feedback @carlocaione  , agreed this change by itself for the new MMIO API should be a standalone PR.

KarthikL1729 commented:
I have dropped those commits from here. Will make a new branch and add RAT and external address translation in a new PR.

KarthikL1729 committed:
soc: ti_k3: Added SoC files for Cortex M4F on ti_am62x_sk

Added SoC support files for the Cortex M4F core on the TI AM62X SK EVM.

Signed-off-by: L Lakshmanan <l-lakshmanan@ti.com>

KarthikL1729 committed:
dts: ti_am62x_sk: Added base devicetree file for AM62X SK

Added the base devicetree file for the TI AM62X SK EVM board.

Signed-off-by: L Lakshmanan <l-lakshmanan@ti.com>

KarthikL1729 committed:
board: ti_am62x_sk_m4: Added board files for TI AM62X SK

Added configuration and documentation files for the AM62x board
M4 core.

Signed-off-by: L Lakshmanan <l-lakshmanan@ti.com>

KarthikL1729 commented:
@carlocaione @vaishnavachath This PR is dependent on #60028 being merged, and it includes all the changes to get the board working using the mm api. Requesting this to be reviewed.

carlocaione commented:
@KarthikL1729 can you make sure that the CI is green?

gramsay0 commented on drivers/serial/uart_ns16550.c:
```diff
@@ -403,6 +407,10 @@ static int uart_ns16550_configure(const struct device *dev,
 	}
 #endif
 
+#ifdef CONFIG_SOC_AM62x_M4
```
Could this be something more generic like `CONFIG_UART_NS16550_16X_MODE`?
And rename `MDR1_UART_MODE` -> `MDR1_UART_16X_MODE`

KarthikL1729 replied:
What is 16X mode? I can't seem to find the config.

gramsay0 replied:
I was suggesting to make a new config for the NS16550 UART.
This UART driver is used by many different SOCs.
Avoiding specific configs (`CONFIG_SOC_AM62x_M4`) in here will help keep this file tidy and reduce how frequently it is modified.

> What is 16X mode? 

I think that is the mode you are selecting in the MDR1 register -> 16X oversampling

Maybe one of the code owners could comment on how to best do this? @dcpleung @nashif @aasthagr

KarthikL1729 replied:
I will open a new PR for this, and add a variant for TI K3, as there are more registers for modes.

gramsay0 commented on soc/arm/ti_k3/am62x_m4/linker.ld:
```diff
@@ -0,0 +1,19 @@
+	SECTION_PROLOGUE(.resource_table,, SUBALIGN(4))
+	{
+		KEEP(*(.resource_table*))
+    } GROUP_LINK_IN(DDR)
```
Tidy indent

gramsay0 commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -27,10 +35,35 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintp
+
+	baseAddr = (uint32_t) virt_reg_base;
+	/* Lock 0 */
+	kickAddr = (volatile uint32_t *)(baseAddr + CSL_MCU_PADCONFIG_LOCK0_KICK0_OFFSET);
```
Consider making `kickAddr` a `uintptr_t` to avoid some of the typecasting

gramsay0 commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -27,10 +35,35 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintp
 	const struct device *dev = DEVICE_DT_GET(PINCTRL_NODE);
 	uintptr_t virt_reg_base = DEVICE_MMIO_GET(dev);
 
+	#ifdef CONFIG_SOC_AM62x_M4
```
If this unlocking just for pinctrl?
If it is needed for other peripherals in the future, maybe create a reusable function in soc.h

KarthikL1729 replied:
It is only required once.

gramsay0 replied:
Move it to a `SYS_INIT()` call if it is only required once.
Currently it may be called multiple times

If it is not related to pinctrl and only needs calling once, then it could be done in a `SYS_INIT()` in soc.c

KarthikL1729 replied:
Have added it to `SYS_INIT()`

gramsay0 commented on boards/arm/ti_am62x_sk_m4/ti_am62x_sk_m4_defconfig:
```diff
@@ -0,0 +1,33 @@
+CONFIG_XIP=n
+
+# Enable RAT module
+CONFIG_EXTERNAL_ADDRESS_TRANSLATION=y
```
Could these be the defaults for the soc series?
Same for `CONFIG_UART_NS16550_ACCESS_WORD_ONLY`

KarthikL1729 replied:
Will add them to the defaults!

gramsay0 reviewed (commented):
This is looking really good, great work!

gramsay0 commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -27,10 +35,35 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintp
 	const struct device *dev = DEVICE_DT_GET(PINCTRL_NODE);
 	uintptr_t virt_reg_base = DEVICE_MMIO_GET(dev);
 
+	#ifdef CONFIG_SOC_AM62x_M4
```
Does this unlocking need to be done on every call, or could it be done once during init?

KarthikL1729 replied:
It should be done only once during init the way I understand it

KarthikL1729 commented:
@carlocaione I appear to have used a few casts in #60028 that are throwing warnings in the build and I think that is why twister build 3 is failing.

I am very sorry about this, it seems to have slipped my mind, and I have fixed the casts locally now, but how do I go about pushing the changes? Should I add them here?

carlocaione commented:
> I am very sorry about this, it seems to have slipped my mind, and I have fixed the casts locally now, but how do I go about pushing the changes? Should I add them here?

nope, open a new standalone PR with the fixes.

KarthikL1729 commented:
> > I am very sorry about this, it seems to have slipped my mind, and I have fixed the casts locally now, but how do I go about pushing the changes? Should I add them here?
> 
> nope, open a new standalone PR with the fixes.

Will do that asap. Thank you so much for considering.

gramsay0 commented on dts/arm/ti/am62x_sk_m4.dtsi:
```diff
@@ -0,0 +1,63 @@
+
+	pinctrl: pinctrl@4080000 {
+		compatible = "ti,k3-pinctrl";
+		reg = <0x04080000 0x8000>;
```
The datasheet defines `MCU_PADCONFIG` as starting from `0x04084000`.
Other projects (like Linux and U-Boot) use `0x04084000 0x88`, could this be kept consistent with those device trees?

KarthikL1729 replied:
I have fixed this. It is now in accordance with the Linux DT. Sorry for the confusion.

vaishnavachath commented:
> @carlocaione @vaishnavachath This PR is dependent on #60028 being merged, and it includes all the changes to get the board working using the mm api. Requesting this to be reviewed.

@KarthikL1729 Thanks, there are still some pending comments that needs to be addressed from previous reviews (removing MMR unlock from pinctrl driver, moving ns16550 changes to separate commit and making it generic .etc), can you please address those and also the newer review comments?

gramsay0 commented on soc/arm/ti_k3/am62x_m4/soc.c:
```diff
@@ -0,0 +1,70 @@
+ */
+};
+
+void mmr_unlock(void)
```
This can be `static`

gramsay0 commented on soc/arm/ti_k3/am62x_m4/soc.c:
```diff
@@ -0,0 +1,70 @@
+#define ADDR_TRANSLATE_REGIONS			(4u)
+#define PINCTRL_BASE_ADDR					(0x4080000u)
+
+struct address_trans_region_config region_config[ADDR_TRANSLATE_REGIONS] = {
```
This can be `static`

gramsay0 commented on soc/arm/ti_k3/am62x_m4/soc.c:
```diff
@@ -0,0 +1,70 @@
+#include <soc.h>
+
+#define ADDR_TRANSLATE_RAT_BASE_ADDR		(0x044200000u)
+#define ADDR_TRANSLATE_REGIONS			(4u)
```
nitpick: This doesn't really need to be defined.
If you remove the size from the array definition, the initialiser list defines the size:
`struct address_trans_region_config region_config[] =  {`
Then you can use `ARRAY_SIZE(region_config)` in the call to `sys_mm_drv_ti_rat_init()`.
This will help prevent `ADDR_TRANSLATE_REGIONS` from being a different size to the initialiser list

gramsay0 commented on soc/arm/ti_k3/am62x_m4/soc.c:
```diff
@@ -0,0 +1,70 @@
+static int am62x_m4_init(void)
+{
+	sys_mm_drv_ti_rat_init(
+		&region_config, ADDR_TRANSLATE_RAT_BASE_ADDR, ADDR_TRANSLATE_REGIONS);
```
Remove unnecessary `&`

KarthikL1729 replied:
I need to pass a pointer to the struct, and region_config is the object, so I think the `&` is required?

gramsay0 replied:
`region_config` is an array (not a struct).
In this case `region_config`, `&region_config` and `&region_config[0]` would all produce the same address.
AFAIK `region_config` is the more standard way to pass an array like this

KarthikL1729 replied:
Oh right, I didn't realise it was an array, will change.

gramsay0 commented on soc/arm/ti_k3/pinctrl_soc.h:
```diff
@@ -0,0 +1,39 @@
+#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop)		\
+	{ DT_FOREACH_PROP_ELEM(node_id, prop, Z_PINCTRL_STATE_PIN_INIT) }
+
+void mmr_unlock(void);
```
Remove prototype

gramsay0 replied:
Remove unused prototype

gramsay0 commented on soc/arm/ti_k3/am62x_m4/soc.h:
```diff
@@ -0,0 +1,14 @@
+
+#include <zephyr/drivers/mm/rat.h>
+
+#define KICK0_UNLOCK_VAL			(0x68EF3490U)
```
These defines could be in soc.h
And remove `mmr_unlock` prototype

gramsay0 reviewed (commented):
Just some very minor tidy up

gramsay0 commented:
@KarthikL1729 I tested this on a Phytec AM62x board and it worked!!!

Shell module sample:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/124088381/1c3e0549-68ba-48c2-a2d9-d498b3eeb1ac)

I noticed these two issues:

1. The build generated a **2.6GB** `build/zephyr/zephyr.bin`
2. The shell module example tries to build with interrupt driven UART which is not yet supported.
It would be nice if this was disabled by default for this board/SoC.
(I had to disable these: `CONFIG_UART_INTERRUPT_DRIVEN=n` `CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN=n`).

KarthikL1729 commented:
> @KarthikL1729 I tested this on a Phytec AM62x board and it worked!!!
> 
> Shell module sample: ![image](https://user-images.githubusercontent.com/124088381/252819469-1c3e0549-68ba-48c2-a2d9-d498b3eeb1ac.png)
> 
> Although I also noticed these two issues:
> 
> 1. The build generated a **2.6GB** `build/zephyr/zephyr.bin`
> 2. The shell module sample tries to build with interrupt driven UART which is not yet supported.
>    It would be nice if this was disabled by default for this board/SoC (until it is supported).
>    (I had to disable these: `CONFIG_UART_INTERRUPT_DRIVEN` `CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN`).

I just noticed this locally as well, the .bin file does appear to be around 2.6GB. The elf file is around 370kB though. 

The reason this seems to be happening is because of the linker. In the Linker, I am placing the resource table in the DDR region (`0x9CCOOOOO`) and there is a large gap from the internal memory to this section. I think the .bin cannot have sections and must be contiguous, so it fills the gaps with zeroes, leading to a very large .bin, while the elf is as expected.

gramsay0 commented:
> > @KarthikL1729 I tested this on a Phytec AM62x board and it worked!!!
> > Shell module sample: ![image](https://user-images.githubusercontent.com/124088381/252819469-1c3e0549-68ba-48c2-a2d9-d498b3eeb1ac.png)
> > Although I also noticed these two issues:
> > 
> > 1. The build generated a **2.6GB** `build/zephyr/zephyr.bin`
> > 2. The shell module sample tries to build with interrupt driven UART which is not yet supported.
> >    It would be nice if this was disabled by default for this board/SoC (until it is supported).
> >    (I had to disable these: `CONFIG_UART_INTERRUPT_DRIVEN` `CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN`).
> 
> I just noticed this locally as well, the .bin file does appear to be around 2.6GB. The elf file is around 370kB though.
> 
> The reason this seems to be happening is because of the linker. In the Linker, I am placing the resource table in the DDR region (`0x9CCOOOOO`) and there is a large gap from the internal memory to this section. I think the .bin cannot have sections and must be contiguous, so it fills the gaps with zeroes, leading to a very large .bin, while the elf is as expected.

Yes, I agree.
I think you should disable the binary output file (`CONFIG_BUILD_OUTPUT_BIN`) for this board/SoC, with a comment explaining this

vaishnavachath commented:
Removing DNM and RFC labels as the primary dependency (RAT support) is merged, @KarthikL1729 once you have the serial driver PR please mention it as a dependency here.

vaishnavachath commented on boards/arm/ti_am62x_sk_m4/doc/index.rst:
```diff
@@ -0,0 +1,147 @@
+
+This board configuration uses a single serial communication channel with the
+MCU domain UART (MCU_UART0).
+
```
MCU UART0 power domains and clocks are setup by bootloader now right? I think we should mention this somewhere in the documentation.

vaishnavachath commented on boards/arm/ti_am62x_sk_m4/ti_am62x_sk_m4_defconfig:
```diff
@@ -0,0 +1,27 @@
+CONFIG_BOARD_TI_AM62X_SK_M4=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_SEMIHOST=y
+
```
SEMIHOST config can be dropped from board defconfig?

KarthikL1729 commented:
> > > @KarthikL1729 I tested this on a Phytec AM62x board and it worked!!!
> > > Shell module sample: ![image](https://user-images.githubusercontent.com/124088381/252819469-1c3e0549-68ba-48c2-a2d9-d498b3eeb1ac.png)
> > > Although I also noticed these two issues:
> > > 
> > > 1. The build generated a **2.6GB** `build/zephyr/zephyr.bin`
> > > 2. The shell module sample tries to build with interrupt driven UART which is not yet supported.
> > >    It would be nice if this was disabled by default for this board/SoC (until it is supported).
> > >    (I had to disable these: `CONFIG_UART_INTERRUPT_DRIVEN` `CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN`).
> > 
> > 
> > I just noticed this locally as well, the .bin file does appear to be around 2.6GB. The elf file is around 370kB though.
> > The reason this seems to be happening is because of the linker. In the Linker, I am placing the resource table in the DDR region (`0x9CCOOOOO`) and there is a large gap from the internal memory to this section. I think the .bin cannot have sections and must be contiguous, so it fills the gaps with zeroes, leading to a very large .bin, while the elf is as expected.
> 
> Yes, I agree. I think you should disable the binary output file (`CONFIG_BUILD_OUTPUT_BIN`) for this board/SoC, with a comment explaining this

Added this as an SoC config and can confirm, no binary generated now. Addressed all other changes as well.

KarthikL1729 commented:
This PR now requires #60325 and #60336 as dependencies to be fully functional.

DhruvaG2000 commented on boards/arm/ti_am62x_sk_m4/doc/index.rst:
```diff
@@ -0,0 +1,148 @@
+
+References
+**********
+
```
just wondering, when github renders this RST the following section comes up blank. Have you made sure the below reference list acually appears when you build zephyr docs?

KarthikL1729 replied:
They are links in the text, however added TRM link now.

DhruvaG2000 commented on soc/arm/ti_k3/Kconfig:
```diff
@@ -0,0 +1,19 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FAMILY_TI_K3
+	bool
```
please add some info in front of the bool

gramsay0 replied:
I think these are intentionally hidden/promptless configs:
https://docs.zephyrproject.org/latest/build/kconfig/tips.html#assignments-to-promptless-symbols-in-configuration-files
Most `config SOC_FAMILY_*` / `config SOC_PART_NUMBER_*` configs follow this pattern

DhruvaG2000 commented on soc/arm/ti_k3/am62x_m4/Kconfig.soc:
```diff
@@ -0,0 +1,28 @@
+endchoice
+
+config SOC_PART_NUMBER_AM62x
+	bool
```
ditto - add some bool "foo"

DhruvaG2000 commented on soc/arm/ti_k3/am62x_m4/soc.c:
```diff
@@ -0,0 +1,73 @@
+#include <soc.h>
+
+#define ADDR_TRANSLATE_RAT_BASE_ADDR		(0x044200000u)
+#define PINCTRL_BASE_ADDR					(0x4080000u)
```
alignment here looks a bit weird?
I think someone else pointed this out as well, not sure.

KarthikL1729 replied:
Fixed.

vaishnavachath commented on boards/arm/ti_am62x_sk_m4/doc/index.rst:
```diff
@@ -0,0 +1,148 @@
+   http://www.ti.com/tool/ccstudio
+
+.. _CCS Debug Handbook:
+   http://processors.wiki.ti.com/index.php/Debug_Handbook_for_CCS#Cortex_M3.2FM4.2FTiva.2FMSP432
```
Other links are relevant, but this one seems to be for MSP432? CCS setup page also is not needed? for more information there is mcu-plus-sdk page already provided, looks like these are here since originally you had only CCS based loading support.

KarthikL1729 replied:
Removed all unnecessary links.

DhruvaG2000 reviewed (approved):
Thanks for addressing all my concerns, and good job on finally adding zephyr support for M4 on the TI AM62x!

Tested on my SK-AM625 as well 😄  
![image](https://github.com/zephyrproject-rtos/zephyr/assets/41233856/d9af2f45-1430-4089-998a-82f2cbfc36f5)

carlocaione reviewed (dismissed):
Good job folks!

vaishnavachath commented on dts/arm/ti/am62x_sk_m4.dtsi:
```diff
@@ -0,0 +1,63 @@
+		clock-frequency = <48000000>;
+		current-speed = <115200>;
+		reg-shift = <2>;
+		status = "okay";
```
disable this in SoC dtsi, enable in sk_m4 dts?

KarthikL1729 commented:
Requesting this PR to not be merged, as I am working on a fix.

KarthikL1729 commented:
I have added the fix, and the twister build (3) test only fails due to an unused variable k in the virtio driver I think. #59191 faces the same problem, and I think there it was discussed that #60375 fixes that issue. So I think this PR can be merged now

It has been tested on the board as well.

gramsay0 commented on dts/arm/ti/am62x_sk_m4.dtsi:
```diff
@@ -0,0 +1,65 @@
+	uart0: serial@4a00000 {
+		compatible = "ns16550";
+		reg = <0x04a00000 0x200>;
+		interrupts = <40 4>;
```
The interrupt should be 24:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/124088381/f03da761-5b51-4be3-81dd-be2d05c5ba05)
I tested this with `interrupts = <24 4>;` and the shell module sample with ISR driver UARTs works!

gramsay0 replied:
BTW thanks for fixing this properly, rather than the quick fix (disabling UART ISRs) I had suggested

KarthikL1729 replied:
Fixed this. Thank you for the review!

gramsay0 commented on soc/arm/ti_k3/am62x_m4/soc.c:
```diff
@@ -0,0 +1,73 @@
+#define CSL_MCU_PADCONFIG_LOCK0_KICK0_OFFSET	(0x1008)
+#define CSL_MCU_PADCONFIG_LOCK1_KICK0_OFFSET	(0x5008)
+
+struct address_trans_region_config region_config[] = {
```
`static struct`
